core/server/master/src/main/java/alluxio/master/file/meta/InvalidationSyncCache.java

```diff
@@ -0,0 +1,413 @@
+              if (removal.wasEvicted() && removal.getKey() != null && removal.getValue() != null) {
+                if (onRemoval != null) {
+                  onRemoval.accept((String) removal.getKey(), (SyncState) removal.getValue());
+                }
+                onCacheEviction((String) removal.getKey(), (SyncState) removal.getValue());
+              }
+            })
+        .maximumSize(Configuration.getInt(
+            PropertyKey.MASTER_UFS_PATH_CACHE_CAPACITY))
+        .build();
+  }
+
+  /**
+   * Called when an element is evicted from the cache.
+   * @param path the path
+   * @param state the state
+   */
+  void onCacheEviction(String path, SyncState state) {
+    try {
+      // on eviction, we must mark our parent as needing a sync with our invalidation time
+      notifyInvalidationInternal(PathUtils.getParent(path), state.mInvalidationTime);
+    } catch (InvalidPathException e) {
+      throw new RuntimeException("Should not have an invalid path in the cache", e);
+    }
+  }
+
+  /**
+   * There is a most one writer at a time to a sync state.
+   * But multiple readers may be reading the values while they
+   * are being updated, thus the values are volatile.
+   */
+  static class SyncState {
+    volatile long mDirectChildInvalidation = 0;
+    volatile long mRecursiveChildInvalidation = 0;
+    volatile long mInvalidationTime = 0;
+    volatile long mValidationTime = 0;
+    volatile long mDirectValidationTime = 0;
+    volatile long mRecursiveValidationTime = 0;
+    volatile boolean mIsFile;
+
+    SyncState(boolean isFile) {
+      mIsFile = isFile;
+    }
+
+    void setInvalidationTime(long time) {
+      if (time > mInvalidationTime) {
+        mInvalidationTime = time;
+      }
+    }
+
+    void setDirectChildInvalidation(long time) {
+      if (time > mDirectChildInvalidation) {
+        mDirectChildInvalidation = time;
+      }
+    }
+
+    void setRecursiveChildInvalidation(long time) {
+      if (time > mRecursiveChildInvalidation) {
+        mRecursiveChildInvalidation = time;
+      }
+    }
+
+    void setIsFile(boolean isFile) {
+      // only transition from file to directory
+      // and not the other way around
+      if (!isFile && mIsFile) {
+        mIsFile = false;
+      }
+    }
+
+    void setValidationTime(long time, DescendantType descendantType) {
+      if (time > mValidationTime) {
+        mValidationTime = time;
+      }
+      if (descendantType == DescendantType.ALL) {
+        if (time > mRecursiveValidationTime) {
+          mRecursiveValidationTime = time;
+        }
+        if (time > mDirectValidationTime) {
+          mDirectValidationTime = time;
+        }
+      } else if (descendantType == DescendantType.ONE) {
+        if (time > mDirectValidationTime) {
+          mDirectValidationTime = time;
+        }
+      }
+    }
+  }
+
+  /**
+   * Called when starting a sync.
+   * @return the time at the start of the sync
+   */
+  public long startSync() {
```

Comment:
would recordSyncStart be a better naming?
you don't actually start a sync here.